File: customer_analytics_adaptive_upgrade/frontend/services/adaptive.ts
import { apiClient } from '@/services/api'

export type TargetCandidate = { column: string; confidence: number; kind: string; values: string[]; reason: string }
export type ColumnProfile = { name: string; inferredType: string; semanticRoles: string[]; missingCount: number; missingRate: number; uniqueCount: number; sampleValues: string[]; warnings: string[] }
export type Dataset = { id: string; projectId: string; originalName: string; rowCount: number; columnCount: number; profile: { columns: ColumnProfile[]; identifierCandidates: Array<{ column: string; confidence: number; reason: string }>; targetCandidates: TargetCandidate[]; duplicateRows: number }; target: string | null; problemType: string }
export type Project = { id: string; name: string; description: string; createdAt: string; updatedAt: string; datasetCount: number }
export type ModelRun = { id: string; algorithm: string; problemType: string; productionThreshold: number | null; metrics: Record<string, number> }

const base = '/adaptive/projects'

export const adaptiveApi = {
  listProjects: () => apiClient.get<Project[]>(base),
  createProject: (name: string, description?: string) => apiClient.post<Project>(base, { name, description }),
  uploadDataset: (projectId: string, file: File) => {
    const form = new FormData()
    form.append('file', file)
    return apiClient.post<Dataset>(`${base}/${projectId}/datasets`, form, { headers: { 'Content-Type': 'multipart/form-data' } })
  },
  getDataset: (projectId: string, datasetId: string) => apiClient.get<Dataset>(`${base}/${projectId}/datasets/${datasetId}`),
  confirmTarget: (projectId: string, datasetId: string, targetColumn: string) => apiClient.post<Dataset>(`${base}/${projectId}/datasets/${datasetId}/target`, { targetColumn }),
  train: (projectId: string, datasetId: string) => apiClient.post<ModelRun>(`${base}/${projectId}/datasets/${datasetId}/train`),
  customers: (projectId: string, datasetId: string, search = '') => apiClient.get<{ identifierColumn: string | null; customers: Array<{ customerKey: string; record: Record<string, unknown> }> }>(`${base}/${projectId}/datasets/${datasetId}/customers`, { params: { search } }),
  predict: (projectId: string, datasetId: string, customerKey: string) => apiClient.post<{ probability?: number; threshold?: number; predictedClass?: string; risk?: string; customerKey: string; record: Record<string, unknown> }>(`${base}/${projectId}/datasets/${datasetId}/predict`, { customerKey }),
  explainLocal: (projectId: string, datasetId: string, customerKey: string) => apiClient.post<{ contributions: Array<{ feature: string; shapValue: number }>; statement: string }>(`${base}/${projectId}/datasets/${datasetId}/explain/local`, { customerKey }),
  explainGlobal: (projectId: string, datasetId: string) => apiClient.get<{ features: Array<{ feature: string; meanAbsoluteShap: number }> }>(`${base}/${projectId}/datasets/${datasetId}/explain/global`),
  listExplanations: (projectId: string, datasetId: string) => apiClient.get<Array<{ id: string; scope: string; customerKey?: string; generatedAt: string }>>(`${base}/${projectId}/datasets/${datasetId}/explanations`),
  recommend: (projectId: string, datasetId: string, customerKey: string) => apiClient.post<{ prediction: { probability: number; threshold: number }; recommendation: { ruleKey: string; recommendationText: string; reason: string; drivers: string[] } }>(`${base}/${projectId}/datasets/${datasetId}/recommend`, { customerKey }),
  listRecommendations: (projectId: string, datasetId: string) => apiClient.get<Array<{ customerKey: string; recommendation: { ruleKey: string; recommendationText: string; reason: string }; generatedAt: string }>>(`${base}/${projectId}/datasets/${datasetId}/recommendations`),
  audit: (projectId: string, datasetId: string) => apiClient.post<{ status: string; checks: Array<{ key: string; status: string; detail: string }> }>(`${base}/${projectId}/datasets/${datasetId}/audit`),
}

File: customer_analytics_adaptive_upgrade/src/api/adaptive.py
from __future__ import annotations

from fastapi import APIRouter, File, HTTPException, Query, UploadFile
from pydantic import BaseModel, Field

from src.adaptive.platform import AdaptiveError, AdaptivePlatform

router = APIRouter(prefix="/api/adaptive", tags=["adaptive-customer-intelligence"])
platform = AdaptivePlatform()


class ProjectInput(BaseModel):
    name: str = Field(min_length=3, max_length=120)
    description: str | None = Field(default=None, max_length=500)


class TargetInput(BaseModel):
    targetColumn: str = Field(min_length=1, max_length=200)


class CustomerInput(BaseModel):
    customerKey: str = Field(min_length=1, max_length=300)
    modelId: str | None = None


def operation(callable):
    try:
        return callable()
    except AdaptiveError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/projects")
def list_projects():
    return platform.list_projects()


@router.post("/projects")
def create_project(payload: ProjectInput):
    return operation(lambda: platform.create_project(payload.name, payload.description))


@router.post("/projects/{project_id}/datasets")
async def upload_dataset(project_id: str, file: UploadFile = File(...)):
    content = await file.read()
    return operation(lambda: platform.upload_dataset(project_id, file.filename or "dataset.csv", content))


@router.get("/projects/{project_id}/datasets/{dataset_id}")
def get_dataset(project_id: str, dataset_id: str):
    return operation(lambda: platform.get_dataset(project_id, dataset_id))


@router.post("/projects/{project_id}/datasets/{dataset_id}/target")
def confirm_target(project_id: str, dataset_id: str, payload: TargetInput):
    return operation(lambda: platform.confirm_target(project_id, dataset_id, payload.targetColumn))


@router.post("/projects/{project_id}/datasets/{dataset_id}/train")
def train(project_id: str, dataset_id: str):
    return operation(lambda: platform.train(project_id, dataset_id))


@router.get("/projects/{project_id}/datasets/{dataset_id}/customers")
def customers(project_id: str, dataset_id: str, search: str = Query(default="", max_length=200)):
    return operation(lambda: platform.list_customers(project_id, dataset_id, search))


@router.post("/projects/{project_id}/datasets/{dataset_id}/predict")
def predict(project_id: str, dataset_id: str, payload: CustomerInput):
    return operation(lambda: platform.predict(project_id, dataset_id, payload.customerKey, payload.modelId))


@router.post("/projects/{project_id}/datasets/{dataset_id}/explain/local")
def local_explain(project_id: str, dataset_id: str, payload: CustomerInput):
    return operation(lambda: platform.explain(project_id, dataset_id, payload.customerKey, payload.modelId))


@router.get("/projects/{project_id}/datasets/{dataset_id}/explain/global")
def global_explain(project_id: str, dataset_id: str, model_id: str | None = None):
    return operation(lambda: platform.explain(project_id, dataset_id, None, model_id))


@router.get("/projects/{project_id}/datasets/{dataset_id}/explanations")
def list_explanations(project_id: str, dataset_id: str):
    return operation(lambda: platform.list_records(project_id, "explanations", dataset_id))


@router.post("/projects/{project_id}/datasets/{dataset_id}/recommend")
def recommend(project_id: str, dataset_id: str, payload: CustomerInput):
    return operation(lambda: platform.recommend(project_id, dataset_id, payload.customerKey, payload.modelId))


@router.get("/projects/{project_id}/datasets/{dataset_id}/recommendations")
def list_recommendations(project_id: str, dataset_id: str):
    return operation(lambda: platform.list_records(project_id, "recommendations", dataset_id))


@router.post("/projects/{project_id}/datasets/{dataset_id}/audit")
def audit(project_id: str, dataset_id: str, model_id: str | None = None):
    return operation(lambda: platform.audit(project_id, dataset_id, model_id))
